### PR TITLE
fix(webdriverio): implement switchFrame for Bidi

### PR DESCRIPTION
**Pull request by christian-bromann**
## Proposed changes

This is a new approach to handle frame switches with Bidi. I want to introduce a new command `switchFrame` (aligned with other commands like `switchWindow`) that simplifies the way user can switch into frames.

Command docs:

Switches the active context to a frame, e.g. an iframe on the page. There are multiple ways you can query a frame
on the page:

  - If given a string it switches to the frame with a matching context id, url or url that contains that string
    ```ts
    // switch to a frame that has the url "https://the-internet.herokuapp.com/iframe"
    await browser.switchFrame('https://the-internet.herokuapp.com/iframe')
    // switch to a frame that has a certain context id
    await browser.switchFrame('A5734774C41F8C91D483BDD4022B2EF3')
    ```

  - If given a WebdriverIO element that references an `iframe` element it will switch to that frame
    ```ts
    // switch to a frame element queried from current context
    await browser.switchFrame($('iframe'))
    ```

  - If given a function it will loop through all iframes on the page and call the function with the context
    object. The function should return a boolean indicating if the frame should be selected. Within the function
    you can use WebdriverIO commands to e.g. check if a certain frame element exists.
    ```ts
    // switch to first frame that contains an element with id "#frameContent"
    await browser.switchFrame(() => $('#frameContent').isExisting())
    ```

  - If given `null` it will switch to the top level frame
    ```ts
    // first switch into a frame
    await browser.switchFrame($('iframe'))
    // do more automation within that frame, then ...

    // switch to the top level frame
    await browser.switchFrame(null)
    ```

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Polish (an improvement for an existing feature)
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (improvements to the project's docs)
- [ ] Specification changes (updates to WebDriver command specifications)
- [ ] Internal updates (everything related to internal scripts, governance documentation and CI files)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [ ] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added the necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Backport Request

[//]: # (The current `main` branch is the development branch for WebdriverIO v9. If your change should be released to the current major version of WebdriverIO (v8), please raise another PR with the same changes against the `v8` branch.)

- [ ] This change is solely for `v9` and doesn't need to be back-ported
- [ ] Back-ported PR at `#XXXXX`

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

File: packages/webdriverio/src/commands/browser/switchFrame.ts [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+            ...(context.children || []).map(mapContext)
+        ]
+        const sessionContext = getContextManager(this)
+        const currentContext = await sessionContext.getCurrentContext()
+
+        /**
+         * transform context tree into a flat list of context objects with references
+         * to children
+         */
+        const allContexts: Record<string, FlatContextTree> = tree.contexts.map(mapContext).flat(Infinity)
```
Comment (erwinheitzman) [inline packages/webdriverio/src/commands/browser/switchFrame.ts:106]: what does the .flat(Infinite) do here? haven't seen this before

File: packages/webdriverio/src/commands/browser/switchFrame.ts [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    url: string,
+    contexts: local.BrowsingContextInfoList,
+    matcher: typeof byUrl | typeof byUrlContaining
+): string | undefined {
+    for (const context of contexts) {
+        if (matcher(context, url)) {
+            return context.context
+        }
+
+        if (context.children && context.children.length > 0) {
```
Comment (erwinheitzman) [inline packages/webdriverio/src/commands/browser/switchFrame.ts:184]: not sure if the ? needs to be added between context and children or between children and length or both but it's a bit cleaner
```suggestion
        if (context.children?.length > 0) {
```

File: packages/webdriverio/src/commands/browser/switchFrame.ts [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+/**
+ * Given a certain `BrowsingContextInfoList` and a desired `contextId`, this helper finds
+ * the `BrowsingContextInfo` that matches the `contextId`
+ * @param tree  a context tree returned from `browsingContextGetTree`
+ * @param contextId  a desired context id
+ * @returns the `BrowsingContextInfo` that matches the `contextId` or `undefined` if not found
+ */
+function findContextById (tree: local.BrowsingContextInfoList, contextId: string): local.BrowsingContextInfo | undefined {
```
Comment (erwinheitzman) [inline packages/webdriverio/src/commands/browser/switchFrame.ts:202]: same function as above, could potentially be merged

**Review by erwinheitzman — APPROVED** [review]
Added some suggestions but nothing major, looking great Christian. Well done, love how clean and readable the code is!

File: packages/webdriverio/src/commands/browser/switchFrame.ts (hunk 1 above)
Comment (christian-bromann) [reply packages/webdriverio/src/commands/browser/switchFrame.ts:106]: `.flat()` by default only flattens with a depth of `1`, so:

```ts
[1, [2, [3]]].flat() === [1, 2, [3]]
```

With infinity we flatten the array infinity times:

```ts
[1, [2, [3]]].flat() === [1, 2, 3]
```

File: packages/webdriverio/src/commands/browser/switchFrame.ts (hunk 3 above)
Comment (christian-bromann) [reply packages/webdriverio/src/commands/browser/switchFrame.ts:202]: Good call!

**Comment by danielhjacobs** [comment]
This doesn't appear to work for classic frame elements.

**Pull request merged**